app/controllers/concerns/shopify_app/ensure_authenticated_links.rb

```diff
@@ -10,9 +10,26 @@ module EnsureAuthenticatedLinks
 
     private
 
+    def splash_page
+      splash_page_with_params(shop: current_shopify_domain, host: params[:host])
+    end
+
+    def base_splash_page
+      root_path(return_to: request.fullpath)
+    end
+
+    def splash_page_with_params(params)
+      uri = URI(base_splash_page)
+      uri.query = CGI.parse(uri.query.to_s)
+        .symbolize_keys
+        .transform_values { |v| v.one? ? v.first : v }
+        .merge(params)
+        .to_query
+      uri.to_s
+    end
```

Comment:
I just pushed a simplified version of this. Since we have a predefined set of params here (`return_to`, `shop`, `host`) there's no need to parse the existing query.